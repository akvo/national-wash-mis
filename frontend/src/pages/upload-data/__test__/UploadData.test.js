import { render, fireEvent, waitFor, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { MemoryRouter } from "react-router-dom";
import axios from "axios";
import { act } from "react-dom/test-utils";
import UploadData from "../UploadData";

jest.mock("axios");

describe("page/upload-data", () => {
  let file;

  beforeEach(() => {
    file = new File(["(⌐□_□)"], "chucknorris.png", { type: "image/png" });
  });

  test("test if users can upload data", async () => {
    axios.mockResolvedValueOnce({ status: 200 });

    const { container, getByTestId, getByText } = render(<UploadData />, {
      wrapper: MemoryRouter,
    });
    expect(getByText("Data Upload")).toBeInTheDocument();

    // Checkbox
    expect(getByText("Update Existing Data")).toBeInTheDocument();
    const checkbox = getByTestId("update-existing-data");
    expect(checkbox.checked).toEqual(false);
    fireEvent.click(checkbox);
    expect(checkbox.checked).toEqual(true);

    // Download data
    expect(
      getByText("If you do not already have a template, please download it")
    ).toBeInTheDocument();
    const select = getByTestId("select-form");
    fireEvent.change(
      select.querySelector(".ant-select-selection-search-input"),
      { target: { value: 1 } }
    );

    // Download button
    const templateLink = (
      <a
        href="blob:http://localhost:3000/dbe88d95-638b-46f9-8bf0-abb231f49bc8"
        download="563350033-WASH-in-Schools.xlsx"
      ></a>
    );
    const downloadTemplateBtn = getByTestId("download-template-btn");
    await act(async () => {
      fireEvent.click(downloadTemplateBtn);
      axios.mockResolvedValueOnce({ status: 200, data: templateLink });
    });

    // Drag and drop file
    const uploadedFile = {
      file: {
        uid: "rc-upload-1658902625996-2",
        lastModified: 1654609229865,
        lastModifiedDate: "2022-06-07T13:40:29.865Z",
        name: "7a13eac4-bfd9-46ab-8bb1-c8ea25bd05ed.xlsx",
        size: 9104,
        type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        percent: 0,
        originFileObj: {
          uid: "rc-upload-1658902625996-2",
        },
        status: "uploading",
      },
      fileList: [
        {
          uid: "rc-upload-1658902625996-2",
          lastModified: 1654609229865,
          lastModifiedDate: "2022-06-07T13:40:29.865Z",
          name: "7a13eac4-bfd9-46ab-8bb1-c8ea25bd05ed.xlsx",
          size: 9104,
          type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
          percent: 0,
          originFileObj: {
            uid: "rc-upload-1658902625996-2",
          },
          status: "uploading",
        },
      ],
    };
    const uploader = screen.getByTestId("drop-file");
    await waitFor(() => {
      fireEvent.change(uploader, {
        target: { files: [file] },
      });
      axios.mockResolvedValueOnce({ status: 200, data: uploadedFile });
    });
    expect(uploadedFile).toHaveProperty("file");
    expect(uploadedFile).toHaveProperty("fileList");
    expect(uploadedFile).toMatchSnapshot("uploadedFile");

    expect(container).toMatchSnapshot();
  });
});
