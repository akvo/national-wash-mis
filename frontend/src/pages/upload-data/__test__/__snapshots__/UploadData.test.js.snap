// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`page/upload-data test if users can upload data 1`] = `
<div>
  <div
    id="uploadData"
  >
    <div
      class="ant-row ant-row-space-between"
    >
      <div
        class="ant-col"
      >
        <div
          class="ant-breadcrumb"
        >
          <span>
            <span
              class="ant-breadcrumb-link"
            >
              <a
                href="/control-center"
              >
                <h1
                  class="ant-typography"
                  style="display: inline;"
                >
                  Control Center
                </h1>
              </a>
            </span>
            <span
              class="ant-breadcrumb-separator"
            >
              &gt;
            </span>
          </span>
          <span>
            <span
              class="ant-breadcrumb-link"
            >
              <h2
                class="ant-typography"
                style="display: inline;"
              >
                Data Upload
              </h2>
            </span>
            <span
              class="ant-breadcrumb-separator"
            >
              &gt;
            </span>
          </span>
        </div>
        <div
          class="description-paragraph"
        >
          This is where you :
          <ul>
            <li>
              Download upload template
            </li>
            <li>
              Bulk upload new data
            </li>
            <li>
              Bulk update existing data
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="ant-divider ant-divider-horizontal"
      role="separator"
    />
    <div
      class="ant-row ant-row-middle"
    >
      <label
        class="ant-checkbox-wrapper ant-checkbox-wrapper-checked"
      >
        <span
          class="ant-checkbox ant-checkbox-checked"
        >
          <input
            class="ant-checkbox-input"
            data-testid="update-existing-data"
            id="updateExisting"
            type="checkbox"
            value=""
          />
          <span
            class="ant-checkbox-inner"
          />
        </span>
        <span>
          Update Existing Data
        </span>
      </label>
    </div>
    <div
      class="ant-card ant-card-bordered"
      style="padding: 0px; min-height: 40vh;"
    >
      <div
        class="ant-card-body"
        style="padding: 0px;"
      >
        <div
          class="ant-space ant-space-horizontal ant-space-align-center"
        >
          <div
            class="ant-space-item"
            style="margin-right: 32px;"
          >
            <img
              src="/assets/data-download.svg"
            />
          </div>
          <div
            class="ant-space-item"
            style="margin-right: 32px;"
          >
            <p>
              If you do not already have a template, please download it
            </p>
          </div>
          <div
            class="ant-space-item"
            style="margin-right: 32px;"
          >
            <div
              class="ant-select ant-select-single ant-select-show-arrow ant-select-open"
              data-testid="select-form"
            >
              <div
                class="ant-select-selector"
              >
                <span
                  class="ant-select-selection-search"
                >
                  <input
                    aria-activedescendant="rc_select_TEST_OR_SSR_list_-1"
                    aria-autocomplete="list"
                    aria-controls="rc_select_TEST_OR_SSR_list"
                    aria-expanded="true"
                    aria-haspopup="listbox"
                    aria-owns="rc_select_TEST_OR_SSR_list"
                    autocomplete="off"
                    class="ant-select-selection-search-input"
                    id="rc_select_TEST_OR_SSR"
                    readonly=""
                    role="combobox"
                    style="opacity: 0;"
                    type="search"
                    unselectable="on"
                    value=""
                  />
                </span>
                <span
                  class="ant-select-selection-placeholder"
                  style="visibility: hidden;"
                >
                  Select Form...
                </span>
              </div>
              <span
                aria-hidden="true"
                class="ant-select-arrow"
                style="user-select: none;"
                unselectable="on"
              >
                <span
                  aria-label="down"
                  class="anticon anticon-down ant-select-suffix"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                  </svg>
                </span>
              </span>
            </div>
          </div>
          <div
            class="ant-space-item"
          >
            <button
              ant-click-animating-without-extra-node="true"
              class="ant-btn ant-btn-primary"
              data-testid="download-template-btn"
              type="button"
            >
              <span>
                Download
              </span>
            </button>
          </div>
        </div>
        <div
          class="ant-space ant-space-horizontal ant-space-align-center"
        >
          <div
            class="ant-space-item"
            style="margin-right: 32px;"
          >
            <img
              src="/assets/data-upload.svg"
            />
          </div>
          <div
            class="ant-space-item"
            style="margin-right: 32px;"
          >
            <p>
              Upload your data
            </p>
          </div>
          <div
            class="ant-space-item"
          >
            <div
              class="ant-select ant-select-single ant-select-show-arrow"
            >
              <div
                class="ant-select-selector"
              >
                <span
                  class="ant-select-selection-search"
                >
                  <input
                    aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                    aria-autocomplete="list"
                    aria-controls="rc_select_TEST_OR_SSR_list"
                    aria-haspopup="listbox"
                    aria-owns="rc_select_TEST_OR_SSR_list"
                    autocomplete="off"
                    class="ant-select-selection-search-input"
                    id="rc_select_TEST_OR_SSR"
                    readonly=""
                    role="combobox"
                    style="opacity: 0;"
                    type="search"
                    unselectable="on"
                    value=""
                  />
                </span>
                <span
                  class="ant-select-selection-item"
                >
                  <span>
                    <div
                      class="ant-upload ant-upload-drag ant-upload-disabled"
                    >
                      <span
                        class="ant-upload ant-upload-disabled ant-upload-btn"
                        role="button"
                      >
                        <input
                          accept="application/vnd.ms-excel,application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
                          data-testid="drop-file"
                          style="display: none;"
                          type="file"
                        />
                        <div
                          class="ant-upload-drag-container"
                        >
                          <p
                            class="ant-upload-drag-icon"
                          >
                            <span
                              aria-label="file-text"
                              class="anticon anticon-file-text"
                              role="img"
                              style="color: rgb(112, 112, 112);"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="file-text"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M854.6 288.7c6 6 9.4 14.1 9.4 22.6V928c0 17.7-14.3 32-32 32H192c-17.7 0-32-14.3-32-32V96c0-17.7 14.3-32 32-32h424.7c8.5 0 16.7 3.4 22.7 9.4l215.2 215.3zM790.2 326L602 137.8V326h188.2zM320 482a8 8 0 00-8 8v48a8 8 0 008 8h384a8 8 0 008-8v-48a8 8 0 00-8-8H320zm0 136a8 8 0 00-8 8v48a8 8 0 008 8h184a8 8 0 008-8v-48a8 8 0 00-8-8H320z"
                                />
                              </svg>
                            </span>
                          </p>
                          <p
                            class="ant-upload-text"
                          >
                            Please select a form
                          </p>
                          <button
                            class="ant-btn"
                            disabled=""
                            type="button"
                          >
                            <span>
                              Browse your computer
                            </span>
                          </button>
                        </div>
                      </span>
                    </div>
                  </span>
                </span>
              </div>
              <span
                aria-hidden="true"
                class="ant-select-arrow"
                style="user-select: none;"
                unselectable="on"
              >
                <span
                  aria-label="down"
                  class="anticon anticon-down ant-select-suffix"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                  </svg>
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`page/upload-data test if users can upload data: uploadedFile 1`] = `
Object {
  "file": Object {
    "lastModified": 1654609229865,
    "lastModifiedDate": "2022-06-07T13:40:29.865Z",
    "name": "7a13eac4-bfd9-46ab-8bb1-c8ea25bd05ed.xlsx",
    "originFileObj": Object {
      "uid": "rc-upload-1658902625996-2",
    },
    "percent": 0,
    "size": 9104,
    "status": "uploading",
    "type": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    "uid": "rc-upload-1658902625996-2",
  },
  "fileList": Array [
    Object {
      "lastModified": 1654609229865,
      "lastModifiedDate": "2022-06-07T13:40:29.865Z",
      "name": "7a13eac4-bfd9-46ab-8bb1-c8ea25bd05ed.xlsx",
      "originFileObj": Object {
        "uid": "rc-upload-1658902625996-2",
      },
      "percent": 0,
      "size": 9104,
      "status": "uploading",
      "type": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
      "uid": "rc-upload-1658902625996-2",
    },
  ],
}
`;
